packages/components/src/text/stories/index.js

```diff
@@ -0,0 +1,38 @@
+/**
+ * Internal dependencies
+ */
+import { Text } from '../text.styles';
+
+export default {
+	title: 'Components|Text',
+	component: Text,
+};
+
+export const _default = () => (
+	<>
+		<Text variant="title.large" as="h1">Title Large</Text>
```

Comment:
Could you explain the preference to name these as strings, vs. expecting a developer to reference some imported JavaScript object, e.g. `variant={ title.large }` ?

My main fear is that referencing these options may be very unintuitive. We see this already with something like the block category, which is of a set of predefined options. Even as a member of the core team,, I find myself needing to [Google this resource](https://developer.wordpress.org/block-editor/developers/block-api/block-registration/#category) to find what the options are. This is not a very good developer workflow.

Conversely, with tooling in many editors, it might be easier to "discover" these options by object traversal (e.g. [VSCode Intellisense](https://code.visualstudio.com/docs/editor/intellisense)) and/or be warned about options which don't exist.

Then again, maybe the same tooling could be smart enough to compare the prop value against the [`TextVariant` typedef you've declared](https://github.com/WordPress/gutenberg/blob/5ebaee2382bdf169b7cc1e39836ba8c6be792ff5/packages/components/src/text/mixins.js#L81). I see you've included this as a documented prop in the `text` mixin, so perhaps this is what you had in mind.

Reply:
I'm not against a seperate component for each type of component - that was [the initial direction I went as I was prototyping the design experiments](https://salad-ui.now.sh/components/title/).

I changed the approach when contributing the component to `@wordpress/components` because 1) a `Button` component already exists 2) remembering the names of each of the typography styles can be hard whereas remembering the `Text` component and autocompleting its properties is much simpler.

Yep VSCode provides autocompletion for the `variant` prop (using Typescript under the hood). I'm not sure how it fares with other editors. 

![image](https://user-images.githubusercontent.com/2237996/69581665-03b43000-102b-11ea-8116-f1d229d3a42c.png)

**Aside:** Would we consider using Typescript for the `@wordpress/components`? Even if the consumer is using JavaScript, it provides better documentation and intellisense.

Reply:
Oh, just reading your example again. I've considered using enums in previous component systems. (With Typescript) the ability to refactor if any of the enum names change is :amaze: but it is limited to the component codebase and in reality the names don't (or shouldn't) change frequently. Other than the ability to refactor :allthethings: there isn't really any benefit of an enum over a union of strings - they both get autocompletion. The down side to enums is that you have to import an extra object for many props for many components which gets tiresome quickly.

Reply:
> **Aside:** Would we consider using Typescript for `@wordpress/components`? Even if the consumer is using JavaScript, it provides them with better documentation and intellisense.

It's come up a few times in the past. Historically, we tended to avoid it, in order to avoid yet another barrier-to-entry for new contributors. Anecdotally, I've been quite happy with using TypeScript _tooling_ in combination with JavaScript + [JSDoc support](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#supported-jsdoc) in my own projects, so that's an area I'd definitely be interested to push further in Gutenberg. Curious if you think there's anything we couldn't achieve with that approach.

Related: #17014, #15027

Reply:
I guess since `typescript@3.7` supports the use of `allowJs` and `declaration` together that means we can start to generate declarations in order to provide value for consuming developers as well as the value that the jsdoc already provides for the developers of the components themselves.

I haven't used the `jsdoc` only approach in a significant sized codebase yet so I can't comment on the effectiveness of that.

Thanks for linking the issues, will follow.